### PR TITLE
Optimise foreach on builtin::indexed ARRAY

### PR DESCRIPTION
**Pull request by leonerd**
Rather than generating an entire temporary list that is twice as big as the original array, instead set a flag on the `OP_ITER` that tells it to set one of the iteration variables to the current array index and use the same `CXt_LOOP_ARY` optimisastion that regular foreach over an array would use.

 * Consider whether `builtin::indexed LIST...` should do the same optimisation. Likely the answer is "yes", but in a separate change.

* This set of changes requires a perldelta entry, and it is included

### CODE REVIEW

**Comment by leonerd** [comment]
This change does have the potential impact that, before it, any modifications to the iterated-on array that happen during the body of its own `foreach` loop would not get seen, whereas now they would. This is the same issue that single-variable `foreach` on a plain array already had, so it's not new. But it is a change for existing code that already calls this function.

File: op.c [hunk 1]
```diff
@@ -171,6 +171,9 @@ recursive, but it's recursive on basic blocks, not on tree nodes.
 
 static const char array_passed_to_stat[] = "Array passed to stat will be coerced to a scalar";
 
+/* UGH!! */
+extern void XS_builtin_indexed(pTHX_ CV *);
```
Comment (wolfsage) [inline op.c]: (Should this be EXTERN_C if we keep this?)

Comment (leonerd) [reply op.c]: > (Should this be EXTERN_C if we keep this?)

Ah quite possibly. Let me try it...

Comment (leonerd) [reply op.c]: Nope, it seems no effect :(

Comment (wolfsage) [reply op.c]: You sure? https://github.com/Perl/perl5/actions/runs/11152032815/job/30997152972?pr=22641

File: lib/builtin.t [hunk 2]
```diff
@@ -347,6 +347,30 @@ package FetchStoreCounter {
     is(prototype(\&builtin::indexed), '@', 'indexed prototype');
 }
 
+# indexed + foreach loop optimisation appears transparent
```
Comment (tonycoz) [inline lib/builtin.t:350]: commit message "optimisa**s**tion" too many s even for British spelling

File: op.c (hunk 1 above)
Comment (leonerd) [reply op.c]: Huh, I blame bad caching.

File: lib/builtin.t (hunk 2 above)
Comment (leonerd) [reply lib/builtin.t:350]: Ahyes, fixed + forcepushed

File: builtin.c [hunk 3]
```diff
@@ -490,9 +490,12 @@ static OP *ck_builtin_func1(pTHX_ OP *entersubop, GV *namegv, SV *ckobj)
     }
 }
 
-XS(XS_builtin_indexed)
+/* This does not use the XS() macro so that op.c can see its prototype */
+void
+Perl_XS_builtin_indexed(pTHX_ CV *cv)
```
Comment (tonycoz) [inline builtin.c:495]:
```
builtin.c: In function ‘Perl_XS_builtin_indexed’:
builtin.c:495:35: warning: unused parameter ‘cv’ [-Wunused-parameter]
  495 | Perl_XS_builtin_indexed(pTHX_ CV *cv)
      |                               ~~~~^~
```
Normally suppressed by the attribute in :
```
#define XSPROTO(name) void name(pTHX_ CV* cv __attribute__unused__)
```

**Comment by tonycoz** [comment]
Deparse needs an update:
```
$ ./perl -Ilib -MO=Deparse -Mbuiltin=indexed -e 'for my ($i, $v) (indexed @ARGV) { print "$i: $v\n"; }' 1 2 3
use builtin (split(/,/, 'indexed', 0));
foreach my ($i, $v) (@ARGV) {
    print "${i}: $v\n";
}
-e syntax OK
```

**Comment by tonycoz** [comment]
Looks ok otherwise

**Comment by Grinnz** [comment]
> This change does have the potential impact that, before it, any modifications to the iterated-on array that happen during the body of its own `foreach` loop would not get seen, whereas now they would. This is the same issue that single-variable `foreach` on a plain array already had, so it's not new. But it is a change for existing code that already calls this function.

Consistency seems good, and the foreach docs basically say "don't do this" already. But maybe this change should have a brief mention in perldelta, since this feature isnt experimental anymore?

File: builtin.c (hunk 3 above)
Comment (leonerd) [reply builtin.c:495]: Fixed and repushed.

**Comment by leonerd** [comment]
Added some Deparse support and repushed. Should be good now.

**Review by tonycoz — APPROVED** [review]

**Pull request merged**